### PR TITLE
Reduce duplication of dependency information

### PR DESCRIPTION
**Pull request by IRDonch**
### Summary

Currently, all runtime dependencies are listed in both `requirements.txt` and `setup.py`, except that they sometimes have differing version constraints. This is confusing and increases the amount of work a maintainer needs to do when updating the dependency list.

Fix it by moving all runtime dependencies to a separate file (`requirements-core.txt`) and referencing that file from both `setup.py` and `requirements.txt`. Combine all version constraints from both original files.

Unfortunately, some things still need to be duplicated:

* The OpenCV dependency varies depending on environment variables, and that logic can't be replicated with a requirements file. So OpenCV has to be listed separately in `setup.py` and `requirements.txt`.

* Cython is a build dependency, not a runtime dependency, so it shouldn't go into `requirements-core.txt`. It would technically be possible to move it to a separate requirements file and reference from there, but it's just one dependency, so it doesn't seem worth it.

### How to test
Try to run `pip install -r requirements.txt`. (The installation through `pip install .` is covered by CI.)

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- ~~[ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly~~ No update should be needed
- ~~[ ] I have added tests to cover my changes~~
- ~~[ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)~~

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: requirements-core.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+attrs>=19.3.0
+defusedxml>=0.6.0
+GitPython>=3.0.8
+lxml>=4.4.1
+matplotlib>=3.3.1
+numpy>=1.17.3
+Pillow>=6.1.0
+
+# Avoid 2.0.2 Linux binary distribution because of
```
Comment (zhiltsov-max) [inline requirements-core.txt:9]: Technically, in requirements.txt we could just use --no-binary, and I'd prefer this solution. But it is not supported in setup.py.

**Review by zhiltsov-max — COMMENTED** [review]
What do you think about adding a couple of extra requirement files? 
`-testing` with pytest
`-ci` with coverage
`-plugins` with TF, TF-gpu, pandas?

**Review by zhiltsov-max — COMMENTED** [review]
Also, you asked about fixed dependency versions in setup.py - consult this doc https://packaging.python.org/discussions/install-requires-vs-requirements/.

**Comment by IRDonch** [comment]
> What do you think about adding a couple of extra requirement files?
> `-testing` with pytest
> `-ci` with coverage
> `-plugins` with TF, TF-gpu, pandas?

I think the most sensible scheme would be as follows:

* `requirements-core.txt` - core runtime dependencies (the file I created in this PR)
* For each plugin that has additional dependencies, `requirements-<plugin>.txt`. The `setup.py` equivalent would be an entry in `extras-require`.
* `requirements-dev.txt` - development dependencies (pytest, pylint, etc.). I don't think we need to split this into test, coverage or whatever.
* `requirements.txt` - master file that includes all of the above.

That said, the above scheme only makes sense if we want to keep runtime dependencies in `requirements.txt` files, which I'm not sure about. A simpler scheme would be to keep them only in `setup.py` and have only a `requirements-dev.txt` and a master `requirements.txt` file that just contains:

```
-e .[plugin1,plugin2]
-r requirements-dev.txt
```

**Comment by IRDonch** [comment]
> Also, you asked about fixed dependency versions in setup.py

I... don't think I did, though? :-) I don't think `setup.py` should have fixed dependency versions, and the document you linked agrees with me.

**Comment by IRDonch** [comment]
So, what now? Should I change anything?

**Review by zhiltsov-max — APPROVED** [review]

**Comment by zhiltsov-max** [comment]
Do you want to add anything into CONTRIBUTING.md?

**Comment by IRDonch** [comment]
> Do you want to add anything into CONTRIBUTING.md?

Not really. I don't think anything I did here conflicts with anything that `CONTRIBUTING.md` says.

**Pull request merged**